### PR TITLE
Reimplement more builder integration tests as unit tests

### PR DESCRIPTION
**Pull request by tkopczynski**
Reimplement some more builder integration tests as unit tests.

Part of resolving #20784

Signed-off-by: Tomasz Kopczynski tomek@kopczynski.net.pl

### CODE REVIEW

**Comment by LK4D4** [comment]
It's kinda funny because 1.5 years ago I moved builder unit tests to integration tests :)

**Comment by LK4D4** [comment]
LGTM

**Comment by vdemeester** [comment]
LGTM 🐯
@LK4D4 for real ? 😜

**Comment by tkopczynski** [comment]
@LK4D4 @vdemeester but it's still ok to move more integration tests to unit tests ? :)

File: builder/dockerfile/evaluator_test.go [hunk 1]
```diff
@@ -66,6 +144,10 @@ func executeTestCase(t *testing.T, testCase dispatchTestCase) {
 	contextDir, cleanup := createTestTempDir(t, "", "builder-dockerfile-test")
 	defer cleanup()
 
+	for filename, content := range testCase.files {
+		createTestTempFile(t, contextDir, filename, content, 0777)
```
Comment (runcom) [inline builder/dockerfile/evaluator_test.go]: does this need a check on err?

**Comment by runcom** [comment]
Moved back to code review, this is failing on WindowsTP5 CI and seems related (?):
```
19:46:19 --- FAIL: TestEmptyDockerfile (0.00s)
19:46:19 	internals_test.go:38: Error when closing tar context: remove \\?\C:\Users\ContainerAdministrator\AppData\Local\Temp\docker-builder912569372\Dockerfile: The process cannot access the file because it is being used by another process.
19:46:19 FAIL
19:46:19 coverage: 32.2% of statements
```

**Comment by tkopczynski** [comment]
hey @runcom, I think this was a bug in internals.go with not closing opened file properly (or rather not closing it at all in case of an error), ptal

File: builder/dockerfile/internals.go [hunk 2]
```diff
@@ -688,7 +689,6 @@ func (b *Builder) readDockerfile() error {
 		}
 	}
 	b.dockerfile, err = parser.Parse(f)
-	f.Close()
```
Comment (runcom) [inline builder/dockerfile/internals.go]: Nice!

**Comment by runcom** [comment]
LGTM assuming windows CI is green, @LK4D4 PTAL

**Comment by runcom** [comment]
@jhowardmsft could you take a look since the bug was somehow exposed in Windows only

**Comment by vdemeester** [comment]
@runcom failures on windosTP5 is not on `test-unit` though.

**Comment by runcom** [comment]
Yeah, it was before

**Comment by vdemeester** [comment]
```
10:59:41 stderr: fatal: cannot create directory at 'integration-cli': Permission denied
```
Some weirdies happens with windowsTP5

**Comment by runcom** [comment]
windowsTP5 failures seems valid, could you check them out?

**Comment by vdemeester** [comment]
@runcom yeah.. it worries me some integration tests depends on each other :-(

**Comment by jhowardmsft** [comment]
Restarted. Got past unit tests and build integration tests...

**Comment by tkopczynski** [comment]
I see that the latest run is failing as well, unfortunately.
I'll look into this but the errors seem really weird compared to what I've changed :/

**Comment by tkopczynski** [comment]
I think I managed to fix this. It turned out that internals.go needed some further adjustments to get it fully working.
ptal @runcom @vdemeester @jhowardmsft

**Comment by thaJeztah** [comment]
Thanks so much again, @tkopczynski, really appreciated!

**Comment by vdemeester** [comment]
Still LGTM 🐼

**Comment by thaJeztah** [comment]
\o/ :tada:

**Pull request merged**